fdbserver/GlobalTagThrottler.actor.cpp

```diff
@@ -27,197 +27,495 @@
 					tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
 
+					tagsWithQuota.clear();
 					state RangeResult currentQuotas = wait(tr.getRange(tagQuotaKeys, CLIENT_KNOBS->TOO_MANY));
-					TraceEvent("GlobalTagThrottler_ReadCurrentQuotas").detail("Size", currentQuotas.size());
+					TraceEvent("GlobalTagThrottler_ReadCurrentQuotas", self->id).detail("Size", currentQuotas.size());
 					for (auto const kv : currentQuotas) {
 						auto const tag = kv.key.removePrefix(tagQuotaPrefix);
 						auto const quota = ThrottleApi::TagQuotaValue::fromValue(kv.value);
-						self->trackedTags[tag].setQuota(quota);
+						self->tagStatistics[tag].setQuota(quota);
+						tagsWithQuota.insert(tag);
 					}
-
+					self->removeUnseenQuotas(tagsWithQuota);
 					++self->throttledTagChangeId;
-					// FIXME: Should wait on watch instead
-					// wait(tr.watch(tagThrottleSignalKey));
 					wait(delay(5.0));
 					TraceEvent("GlobalTagThrottler_ChangeSignaled");
 					CODE_PROBE(true, "Global tag throttler detected quota changes");
 					break;
 				} catch (Error& e) {
-					TraceEvent("GlobalTagThrottlerMonitoringChangesError", self->id).error(e);
+					TraceEvent("GlobalTagThrottler_MonitoringChangesError", self->id).error(e);
 					wait(tr.onError(e));
 				}
 			}
 		}
 	}
 
 public:
-	GlobalTagThrottlerImpl(Database db, UID id) : db(db), id(id) { traceActor = tracer(this); }
+	GlobalTagThrottlerImpl(Database db, UID id) : db(db), id(id) {}
 	Future<Void> monitorThrottlingChanges() { return monitorThrottlingChanges(this); }
-	void addRequests(TransactionTag tag, int count) { trackedTags[tag].addTransactions(count); }
+	void addRequests(TransactionTag tag, int count) { tagStatistics[tag].addTransactions(count); }
 	uint64_t getThrottledTagChangeId() const { return throttledTagChangeId; }
 	PrioritizedTransactionTagMap<ClientTagThrottleLimits> getClientRates() {
-		// TODO: For now, only enforce total throttling rates.
-		// We should use reserved quotas as well.
 		PrioritizedTransactionTagMap<ClientTagThrottleLimits> result;
-		for (auto& [tag, quotaAndCounters] : trackedTags) {
+		lastBusyReadTagCount = lastBusyWriteTagCount = 0;
+		for (auto& [tag, stats] : tagStatistics) {
 			// Currently there is no differentiation between batch priority and default priority transactions
-			auto const limit = quotaAndCounters.updateAndGetPerClientLimit();
-			if (limit.present()) {
-				result[TransactionPriority::BATCH][tag] = result[TransactionPriority::DEFAULT][tag] = limit.get();
+			auto const readLimitingTps = getLimitingTps(tag, OpType::READ);
+			auto const writeLimitingTps = getLimitingTps(tag, OpType::WRITE);
+			Optional<double> limitingTps;
+			if (readLimitingTps.present() && writeLimitingTps.present()) {
+				limitingTps = std::min(readLimitingTps.get(), writeLimitingTps.get());
+			} else if (readLimitingTps.present()) {
+				limitingTps = readLimitingTps;
+			} else {
+				limitingTps = writeLimitingTps;
+			}
```

Comment:
I think the same logic to handle 2 `Optional<double>` appears at least 4 times. It's better to extract those codes to a common function IMO. 
For example,
```
Optional<double> min(Optional<double> a, Optional<double> b) {
   if (a.present() && b.present()) {
      a = std::min(a.get(), b.get());
   } else if (a.present()) {
      return a;
   } else {
      return b;
   }
}
```

Reply:
That's a good idea, I'll make this change.